### PR TITLE
Bug/weeekends are not ignored in harvest widget

### PR DESCRIPTION
**Pull request by ncreated**
This PR changes default behavior of Harvest widget. By default, weekends are excluded from "last N" days chart.

To have weekends counted in to the stat, specify it directly in configuration file:
```
{
  "name": "harvest",
  "settings": {
    ...
    "includeWeekends": false,
    ...
  }
},
```

--

Make sure these boxes are checked before submitting your Pull Request - thank you!

- [x] [SwiftLint](https://github.com/Realm/SwiftLint) is installed on your system
- [x] The project does not generate any warnings or errors
- [x] The tests are passing
- [x] Branch name is in line with [GitFlow](http://jeffkreeftmeijer.com/2010/why-arent-you-using-git-flow/)

### CODE REVIEW

File: GrandCentralBoard/Widgets/Harvest/HarvestWidgetSettings.swift [hunk 1]
```diff
@@ -25,6 +26,7 @@ struct HarvestWidgetSettings: Decodable {
                                          clientID:        json => "clientID",
                                          clientSecret:    json => "clientSecret",
                                          numberOfDays:    json => "numberOfDays",
+                                         includeWeekends: json =>? "includeWeekends",
```
Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Harvest/HarvestWidgetSettings.swift]: Do you think it would be better to have optional Bool, or to check it like:
`includeWeekends = json =>? "includeWeekends" ?? false`

Cause I think, from the point of view of classes that are using the configuration, it would be better not to see optionals. Because then they are storing the logic about the default value.

File: GrandCentralBoard/Widgets/Harvest/HarvestWidgetBuilder.swift [hunk 2]
```diff
@@ -14,8 +14,16 @@ final class HarvestWidgetBuilder: WidgetBuilding {
...
     func build(json: AnyObject) throws -> WidgetControlling {
         let settings = try HarvestWidgetSettings.decode(json)
-        let harvestAPI = HarvestAPI(account: settings.account, refreshCredentials: settings.refreshCredentials,
-                                     downloader: settings.downloader, numberOfDaysToFetch: settings.numberOfDays)
+        let today = NSDate()
+        let billableDates = BillableDates(referenceDate: today,
+                                          numberOfPreviousDays: settings.numberOfDays,
+                                          includeWeekends: settings.includeWeekends ?? false)
+
+        let harvestAPI = HarvestAPI(account: settings.account,
```
Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Harvest/HarvestWidgetBuilder.swift]: You are providing billable dates only once, during widget creation. So if the widget works for two days, you will still fetch the same (old) days.

**Comment by michallaskowski** [comment]
Looks nice, I just don't know how to treat the time billed on weekends. Maybe we should add it to the time billed by a person in the right chart. 

Like today I billed 4 hours, and I know I need to do 2,5 hours during the weekend to catch up. And then, in the 'Last 6 days' I should be treated as guy who billed a standard amount of time.
If you agree, then the class Billable is still needed, but you need to fetch all the days, and just divide the sum of hours by a number of billable days.

File: GrandCentralBoard/Widgets/Harvest/BillableDates.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    init(referenceDate: NSDate, numberOfPreviousDays: Int, includeWeekends: Bool?) {
+        let calendar = NSCalendar(calendarIdentifier: NSCalendarIdentifierGregorian)!
+        let sunday = 1
+        let saturday = 7
+
+        var previousDays: [NSDate] = []
+        var date = referenceDate
+
+        while previousDays.count < numberOfPreviousDays {
```
Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: Maybe a bit more swifty version would be:
```
dates  = (1...numberOfPreviousDays).flatMap { dayNumberFromToday -> NSDate?
  ...
}

File: GrandCentralBoard/Widgets/Harvest/BillableDates.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            date = date.dateDaysAgo(1)
+
+            if let includeWeekends = includeWeekends where includeWeekends == false {
+                let day = calendar.component(.Weekday, fromDate: date)
+                let isWeekend = (day == sunday || day == saturday)
```
Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: And here is a nice method to check if it is weekend:
https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSCalendar_Class/#//apple_ref/occ/instm/NSCalendar/isDateInWeekend:

(and then you can use locale calendar, instead of gregorian)

Comment (ncreated) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: Didn't know that, thanks! Will do 😉

File: GrandCentralBoard/Widgets/Harvest/HarvestWidgetBuilder.swift (hunk 2 above)
Comment (ncreated) [inline GrandCentralBoard/Widgets/Harvest/HarvestWidgetBuilder.swift]: That's true. Excellent catch! 👍

File: GrandCentralBoard/Widgets/Harvest/BillableDates.swift (hunk 3 above)
Comment (ncreated) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: The problem with using  sequence is that getting the number of elements explicitly is hard. If `includeWeekends` is set to `false` and I use gregorian calendar, the sequence from `1` to `numberOfPreviousDays` won't be sufficient, because some dates will be filtered out as being a weekend. Thus, I'd need to compensate for them, by expanding the sequence before filtering... and so on... And how to deal with other calendars? 😰 

Hence, IMO doing implicit `while` loop is much simpler. Or do you have any good idea for handling `includeWeekends = false` in a sequence style iteration?

Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: Yeah, true. Forgot that you are removing some elements, so you don't know what the sequence should be.

File: GrandCentralBoard/Widgets/Harvest/HarvestWidgetSettings.swift (hunk 1 above)
Comment (ncreated) [inline GrandCentralBoard/Widgets/Harvest/HarvestWidgetSettings.swift]: We had this discussion before with @krris. The above is the convention we started in Bonus and Image widgets.

Although I get your point and it's valid, our motivation was:
- to have only "parsing" responsibility in `*Settings` classes (this way, they're easier for test for different edge cases + no need to maintain default values in tests);
- to put "default" values logic in `WidgetBuilding`, since it receives the config json and configures a widget components, so it sounds like a correct place.

I have no strong feelings one way or the other, it's just a consequence of the choice we made and already implemented in few widgets.

**Comment by ncreated** [comment]
> Looks nice, I just don't know how to treat the time billed on weekends. Maybe we should add it to the time billed by a person in the right chart.
> 
> Like today I billed 4 hours, and I know I need to do 2,5 hours during the weekend to catch up. And then, in the 'Last 6 days' I should be treated as a guy who billed a standard amount of time.
> If you agree, then the class BillableDates is still needed, but you need to fetch all the days, and just divide the sum of hours by a number of billable days

I think we should keep it simple 📏. So we either ignore the weekends or not. With weekends being ignored, all hours billed on Sunday or Saturday get lost. It's a corner case that one works on weekend, and we should not complicate the widget's logic and understanding because of that. WDYT?

**Comment by ncreated** [comment]
@michallaskowski ⚽ 🍩

**Comment by michallaskowski** [comment]
I am not sure, it depends what we want to see on the right chart. I am ok with both options (time divided by working days or all days), since both give some valuable info. So it is good for me as it is.

**Comment by ncreated** [comment]
> it depends what we want to see on the right chart.

Yeah, this is the point. Both approaches have value - one shows if the work balance is fine, second if the team delivers desired amount of hours. Sounds like a business decision 😉.

File: GrandCentralBoard/Widgets/Harvest/BillableDates.swift [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+//  GrandCentralBoard
+//
+//  Created by Maciek Grzybowski on 03.06.2016.
+//  Copyright © 2016 Oktawian Chojnacki. All rights reserved.
+//
+
+import Foundation
+
+
+final class BillableDates {
```
Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: Just one more thing I thought of today. Do we need `BillableDates` class, since it is holding only `[NSDate]`? Maybe you could create an extension on `NSDate` instead, that returns an array?

Comment (ncreated) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: In `BillableDates`, I'm using `NSCalendar`. `NSDate` is just a wrapper over raw timestamp and shouldn't know about the `NSCalendar`. Thus, I don't want do introduce it in extension 👎 
```swift
let _ = NSDate.arrayOfPreviousDays(5, referenceDate: today, includeWeekends: false, calendar: NSCalendar)
```

Comment (michallaskowski) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: Ok, what about discussed extension on `NSCalendar`? :)

Comment (ncreated) [inline GrandCentralBoard/Widgets/Harvest/BillableDates.swift]: Will do 👍

**Comment by ncreated** [comment]
@michallaskowski ⚽

**Comment by michallaskowski** [comment]
Looks great
